sdk/program/src/message/versions.rs

```diff
@@ -0,0 +1,269 @@
+use {
+    crate::{
+        hash::Hash,
+        instruction::CompiledInstruction,
+        message::{v0, Message, MessageHeader},
+        pubkey::Pubkey,
+        sanitize::{Sanitize, SanitizeError},
+        short_vec,
+    },
+    serde::{
+        de::{self, Deserializer, SeqAccess, Visitor},
+        ser::{SerializeTuple, Serializer},
+        {Deserialize, Serialize},
+    },
+    std::fmt,
+};
+
+/// Bit mask that indicates whether a serialized message is versioned.
+pub const MESSAGE_VERSION_PREFIX: u8 = 0x80;
+
+/// Message versions supported by the Solana runtime.
+///
+/// # Serialization
+///
+/// If the first bit is set, the remaining 7 bits will be used to determine
+/// which message version is serialized starting from version `0`. If the first
+/// is bit is not set, all bytes are used to encode the original `Message`
+/// format.
+#[frozen_abi(digest = "5hAeXsHoMFfGREfh8RqJYmAjRcs88sT1YxrpJJq3Ne1f")]
+#[derive(Debug, PartialEq, Eq, Clone, AbiEnumVisitor, AbiExample)]
+pub enum VersionedMessage {
+    Original(Message),
```

Comment:
I found the word "original" confusing as to me "v0" also implies original.  How about renaming "Original" to "Unversioned" or "Legacy"

Reply:
Legacy is better, updated